IdentityCore/src/cache/MSIDTokenCacheKey.m

```diff
@@ -39,4 +47,106 @@ - (id)initWithAccount:(NSString *)account
+            s_adalLibraryString,
+            authority.absoluteString.msidBase64UrlEncode,
+            [self.class getAttributeName:resource.msidBase64UrlEncode],
+            clientId.msidBase64UrlEncode];
+}
+
++ (NSString *)accountWithUserIdentifier:(NSString *)userId
+                            environment:(NSString *)environment
+{
+    return userId? [NSString stringWithFormat:@"%u$%@@%@", s_msalV1, userId, environment]: nil;
+}
+
++ (NSString *)serviceWithAuthority:(NSURL *)authority
+                            scopes:(NSOrderedSet<NSString *> *)scopes
+                          clientId:(NSString *)clientId
+{
+    if (scopes.count == 0)
+    {
+        return nil;
+    }
+        
+    return [NSString stringWithFormat:@"%@$%@$%@",
+            authority? authority.absoluteString.msidBase64UrlEncode : @"",
+            clientId? clientId.msidBase64UrlEncode : @"",
+            scopes? scopes.msidToString.msidBase64UrlEncode : @""];
+}
+
++ (MSIDTokenCacheKey *)keyForAdfsUserTokenWithAuthority:(NSURL *)authority
+                                               clientId:(NSString *)clientId
+                                               resource:(NSString *)resource
+{
+    return [[MSIDTokenCacheKey alloc] initWithAccount:@""
+                                              service:[self.class serviceWithAuthority:authority
+                                                                              resource:resource
+                                                                              clientId:clientId]];
+}
+
+
++ (MSIDTokenCacheKey *)keyWithAuthority:(NSURL *)authority
+                               clientId:(NSString *)clientId
+                               resource:(NSString *)resource
+                                    upn:(NSString *)upn
+{
+    return [[MSIDTokenCacheKey alloc] initWithAccount:upn
+                                              service:[self.class serviceWithAuthority:authority
+                                                                              resource:resource
+                                                                              clientId:clientId]];
+}
+
++ (MSIDTokenCacheKey *)keyForAccessTokenWithAuthority:(NSURL *)authority
+                                             clientId:(NSString *)clientId
+                                               scopes:(NSOrderedSet<NSString *> *)scopes
+                                               userId:(NSString *)userId
+{
+    NSString *service = [self.class serviceWithAuthority:authority scopes:scopes clientId:clientId];
+    NSString *account = [self.class accountWithUserIdentifier:userId environment:authority.msidHostWithPortIfNecessary];
+    
+    return [[MSIDTokenCacheKey alloc] initWithAccount:account service:service];
+}
+
++ (MSIDTokenCacheKey *)keyForAllAccessTokensWithUserId:(NSString *)userId
+                                           environment:(NSString *)environment
+{
+    return nil;
```

Comment:
We should still be able to set the account part?

Reply:
yes most definitely :p